src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs

```diff
@@ -17903,5 +17925,890 @@ public class XAttribute : System.Attribute
+                    }
+
+                    public static void M(IEnumerable<int> values) => throw null;
+                }
+                """;
+
+            var comp = CreateCompilation(source, targetFramework: TargetFramework.Net70);
+            comp.VerifyEmitDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "1 2 3", verify: Verification.FailsPEVerify);
+        }
+
+        [Fact]
+        public void OverloadResolution_ListOfRightTypePreferred()
+        {
+            string source = """
+                using System.Collections.Generic;
+
+                C.M([1, 2, 3]);
+
+                class C
+                {
+                    public static void M(List<int> values)
+                    {
+                        System.Console.Write($"{values[0]} {values[1]} {values[2]}");
+                    }
+
+                    public static void M(List<string> values) => throw null;
+                }
+                """;
+
+            var comp = CreateCompilation(source, targetFramework: TargetFramework.Net70);
+            comp.VerifyEmitDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "1 2 3", verify: Verification.FailsPEVerify);
+        }
+
+        [Fact]
+        public void OverloadResolution_StringVsReadOnlySpanChar()
+        {
+            string source = """
+                using System;
+
+                C.M(['a', 'b', 'c']);
+
+                class C
+                {
+                    public static void M(string s) => throw null;
+
+                    public static void M(ReadOnlySpan<char> s)
+                    {
+                        System.Console.Write($"{s[0]} {s[1]} {s[2]}");
+                    }
+                }
+                """;
+
+            var comp = CreateCompilation(source, targetFramework: TargetFramework.Net70);
+            comp.VerifyEmitDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "a b c", verify: Verification.FailsPEVerify);
+        }
+
+        [Fact]
+        public void OverloadResolution_StringVsSpanInt()
+        {
+            string source = """
+                using System;
+
+                C.M(['a', 'b', 'c']);
+
+                class C
+                {
+                    public static void M(string s) => throw null;
+                    public static void M(ReadOnlySpan<int> s) => throw null;
+                }
+                """;
+
+            var comp = CreateCompilation(source, targetFramework: TargetFramework.Net70);
+            comp.VerifyEmitDiagnostics(
+                // (3,3): error CS0121: The call is ambiguous between the following methods or properties: 'C.M(string)' and 'C.M(ReadOnlySpan<int>)'
+                // C.M(['a', 'b', 'c']);
+                Diagnostic(ErrorCode.ERR_AmbigCall, "M").WithArguments("C.M(string)", "C.M(System.ReadOnlySpan<int>)").WithLocation(3, 3)
+                );
+        }
```

Comment:
> } [](http://example.com/codeflow?start=7&length=2)

Please add tests for cases where the Add() method takes a more strongly-typed argument, to verify that we're not relying on the conversion to the enumerable type.
```csharp
class MyCollection1<T> : IEnumerable
{
    private List<T> _list = new();
    IEnumerator IEnumerable.GetEnumerator() => _list.GetEnumerator();
    public void Add(T t) { _list.Add(t); }
}

class MyCollection2<T, U> : IEnumerable<T> where U : T
{
    private List<T> _list = new();
    IEnumerator IEnumerable.GetEnumerator() => _list.GetEnumerator();
    IEnumerator<T> IEnumerable<T>.GetEnumerator() => _list.GetEnumerator();
    public void Add(U u) { _list.Add(u); }
}

class Program
{
    static void Main()
    {
        MyCollection1<int> x = [1, 2, 3];
        x.Report();
        MyCollection2<object, int> y = [1, 2, 3];
        y.Report();
    }
}
```

Reply:
Consider also testing a case where there are no implicit conversions between the `Add()` argument type and iteration type.
```csharp
class MyCollection : IEnumerable<sbyte>
{
    private List<sbyte> _list = new();
    IEnumerator IEnumerable.GetEnumerator() => _list.GetEnumerator();
    IEnumerator<sbyte> IEnumerable<sbyte>.GetEnumerator() => _list.GetEnumerator();
    public void Add(char c) { _list.Add((sbyte)c); }
}

class Program
{
    static void Main()
    {
        MyCollection x = ['a', 'b', default];
        x.Report();
    }
}
```